### ISSUE TITLE
[Reproduction] Higher Completeness error on DTU compared to pretrained model

### ISSUE THREAD

**Issue by xingchen2022** [JianfeiJ/MonoMVSNet#4]
Hi there,

First of all, thanks for open-sourcing this excellent work!

I am currently trying to retrain and reproduce your results on the DTU dataset. However, I noticed a performance gap between my retrained model and the results reported in your paper/pretrained model. Specifically, while the Accuracy is comparable, the **Completeness Error** is abnormally high, leading to a worse Overall score.

### Training Details
I trained the model on a single **RTX 4090 GPU** using the following command:

```bash
python -u train_dtu.py --logdir $DTU_LOG_DIR --dataset=dtu_yao_vit --batch_size=5 \
--trainpath=$DTU_TRAINING --summary_freq 1000 --resume \
--ndepths 8,8,4,4 --attn_temp 2 --wd 0.0001 --lr 0.001 --epochs 15 \
--lrepochs "10,12,14:2" --trainlist $DTU_TRAINLIST --testlist $DTU_TESTLIST $PY_ARGS \
2>&1 | tee -a $DTU_LOG_DIR/log.txt
```

### Results Comparison
I evaluated the checkpoint `finalmodel_14.ckpt` using the provided `test_dtu_dypcd.py`. Here is the comparison with the pretrained results:

| Metric | My Retraining | Pretrained/Paper |
| :--- | :--- | :--- |
| **Mean Acc** | 0.3180 | 0.3139 |
| **Mean Comp** | **0.2815** | **0.2426** |
| **Overall** | 0.2998 | 0.2782 |

As shown above, the `Mean Comp` (0.2815 vs 0.2426) is significantly higher, which negatively impacts the overall performance.

### Logs for Reference
Here are the test scalars from my last checkpoint:

```json
avg_test_scalars: {
    'loss': 2.003, 
    's0_ce_loss': 0.1596, 
    's1_ce_loss': 0.3525, 
    's2_ce_loss': 0.5362, 
    's3_ce_loss': 0.8425, 
    's3_mono_loss': 0.1124, 
    's3_range_err_ratio': 0.1179,
    'abs_depth_error': 5.2209, 
    'thres2mm_error': 0.1670, 
    'thres4mm_error': 0.1005, 
    'thres8mm_error': 0.0651
}
```

I have also attached the training log curve for the entire procedure below:

<img width="1998" height="874" alt="Image" src="https://github.com/user-attachments/assets/bc22ff48-6dc1-4d1e-a4d9-3b7df2cf3277" />

### Question
Could you provide any insights or suggestions on what might be causing this high completeness error? 

Thanks for your help!

**Comment by JianfeiJ** [comment]
I've also experimented with torch>2.0,  which showed slightly poorer performance, it can't be that bad overall=0.2998.

**Comment by xingchen2022** [comment]
> I've also experimented with torch>2.0, which showed slightly poorer performance, it can't be that bad overall=0.2998.

Got it! Thanks for the insight—that suggests the issue likely isn't related to the PyTorch version.

I will pull a clean version of the code and retrain with `batch_size=4` on my single RTX 4090. 

**Quick question on the Learning Rate:**
Since I am training on a single GPU (whereas the original setup likely used 4), the effective batch size might be different. Do you recommend scaling down the learning rate (e.g., halving it to 0.0005) for this single-GPU setup, or should I stick to the default `0.001`?

---

**Question regarding CPVE Encoding Logic**

Before I start the retraining, I was reading the code related to CVPE encoding and found a snippet that confuses me:

https://github.com/JianfeiJ/MonoMVSNet/blob/1eb99cffc14265d990ae870361c6792067eaffc1/models/FMT_with_CVPE.py#L409
```python
intrinsics = proj_matrices[:, :, 1].clone()
intrinsics = intrinsics[:, :, :3, :3]
intrinsics[:, :, 0, :] *= float(w)   # Question A
intrinsics[:, :, 1, :] *= float(h)

camk = torch.eye(4).view(1, 1, 4, 4).repeat(intrinsics.shape[0], intrinsics.shape[1], 1, 1).to(intrinsics.device).float()
camk[:, :, :3, :3] = intrinsics

extrinsics = proj_matrices[:, :, 0].clone()  # Question B
camk = torch.inverse(camk)

img2world = torch.matmul(extrinsics, camk)
img2world = torch.unbind(img2world, 1)
ref_img2world = img2world[0]
```

I have two doubts regarding the logic here:

  * **A) Intrinsics Scaling:** The input `intrinsics` (from `proj_matrices`) usually appear to be in standard pixel coordinates (unnormalized) in the DTU dataset. If so, multiplying them by `w` and `h` again (`intrinsics[:, :, 0, :] *= float(w)`) would result in incorrect, overly large values. 

  * **B) Coordinate Transformation (w2c vs c2w):** The variable `extrinsics` typically represents the **World-to-Camera (w2c)** transformation matrix $E$. To transform a point from the image plane to world space ($P_{world}$), the standard formula is typically:
   `P_world = inv(E) @ inv(K) @ P_pixel`
    However, the code implements `torch.matmul(extrinsics, camk)`, which corresponds to `$E \cdot K^{-1}$`. Unless `extrinsics` here is already inverted (i.e., Camera-to-World), this operation seems mathematically inconsistent with the variable name `img2world`.

Could you help clarify if this is a specific design choice or a potential implementation bug?

Apologies for the many questions, and thank you very much for your patience and help!

**Comment by JianfeiJ** [comment]
All other training parameters followed the default configuration in the repository, including the learning rate (lr=0.001).

**Comment by xingchen2022** [comment]
> All other training parameters followed the default configuration in the repository, including the learning rate (lr=0.001).

Following up on my previous comment, I have some positive updates.

After reverting the learning rate to 0.001, I obtained reasonable results on the single RTX 4090. Here are the metrics:

> Mean Acc:   0.3044
> Mean Comp:  0.2656
> Overall:    0.2850

Additionally, I previously conducted a separate training run on a single V100 (Batch Size = 5), which yielded consistent performance:

> Mean Acc:   0.3126
> Mean Comp:  0.2564
> Overall:    0.2845

**Summary of Findings:**

**Environment Stability:** Factors such as GPU model (RTX 4090 vs. V100), PyTorch version, and Batch Size do not appear to significantly impact performance.

**Learning Rate Sensitivity:** Once the initial camera parameter logic (CPVE) was fixed, the learning rate proved to be a critical factor. Sticking to the default lr=0.001 is essential for proper convergence.

Thank you again for your patience and for providing the benchmarks that helped narrow this down！💗
 I hope these findings will be helpful to others attempting to reproduce the results!
